### PR TITLE
Add success and failure telemetry events

### PR DESCRIPTION
**Pull request by randomanderson**
This PR add success and failure telemetry events for agent installation. There are 3 main changes:

1.  The `report_telemetry()` function that sends telemetry to the backend
2. Calling `report_telemetry()` in every case there was an `exit`. This consists of:
     a. Setting `ERROR_MESSAGE`
     b. Setting `ERROR_CODE`
     c. Printing the message to the screen
     d. Reporting telemetry
3. Setting `report_failure_url` and `telemetry_url` was moved to earlier in the script to catch more potential errors.

### Testing
It was not possible to test every scenario, but I manually tested the following using the production telemetry intake:
* Successful install
* Successful upgrade
* Failure because of unknown minor revision (set via environment variable)
* Failure because of unknown flavor (set via environment variable)
* Failure because of unknown distribution channel (set via environment variable)
* Failure because of unsupported platform (mac)
* Setting `DD_INSTRUMENTATION_TELEMETRY_ENABLED=false` disables telemetry

### CODE REVIEW

**Comment by randomanderson** [comment]
@DataDog/agent-platform : I'm not able to request reviewers. Can one of you take a look at this?

File: install_script.sh.template
Comment (randomanderson) [inline install_script.sh.template:124]: I will remove this header before merging

**Review by paullegranddc — APPROVED** [review]

File: install_script.sh.template
Comment (KSerrania) [inline install_script.sh.template:123]: How important is it for you that the `$apikey` here is the same as the one used by the Agent? 

The main reason I'm asking this is that the `DD_API_KEY` environment variable passed to the script (which becomes this `$apikey` variable) may not be the one that is set in the `datadog.yaml` file.

For instance, if a `datadog.yaml` config file already exists (due to a prior install), then the `DD_API_KEY` parameter of the script is ignored, and no changes to the `datadog.yaml` file are made, see this code snippet: https://github.com/DataDog/agent-linux-install-script/blob/0a6f442b2182f9a70cd8ed3ea2f50609bb3677c7/install_script.sh.template#L825-L842

In fact, we don't require the `DD_API_KEY` variable at all for upgrades, therefore the `$apikey` may be empty: https://github.com/DataDog/agent-linux-install-script/blob/0a6f442b2182f9a70cd8ed3ea2f50609bb3677c7/install_script.sh.template#L403-L409

File: install_script.sh.template
Comment (KSerrania) [inline install_script.sh.template:282]: Could you make these two sections follow the same format?
```suggestion
report_failure_url="https://api.datadoghq.com/agent_stats/report_failure"
if [ -n "$DD_SITE" ]; then
    report_failure_url="https://api.${DD_SITE}/agent_stats/report_failure"
fi

telemetry_url="https://instrumentation-telemetry-intake.datadoghq.com/api/v2/apmtelemetry"
if [ -n "$DD_SITE" ]; then
    telemetry_url="https://instrumentation-telemetry-intake.${DD_SITE}/api/v2/apmtelemetry"
fi
```

File: install_script.sh.template
Comment (KSerrania) [inline install_script.sh.template:159]: Could the meaning of each error code be made explicit in a comment somewhere in the file? For instance, after a quick reading of the file, it seems like 6 is for input errors, 5 is for OS / packages not supported, 1 is for all other generic errors, etc. and it would be nice to have a recap of these in the file.

File: install_script.sh.template
Comment (KSerrania) [inline install_script.sh.template:804]: Why not use the above `printf` message as the `ERROR_MESSAGE` here, like you did in other places?

File: install_script.sh.template
Comment (yahya-mouman) [reply install_script.sh.template:159]: Yes you are right they are already defined in this sheet here https://docs.google.com/spreadsheets/d/1GWk23Y5VRw0IK7wdhNMy10OZYTwODasI518deYiMM_Y/edit#gid=0

Comment (randomanderson) [reply install_script.sh.template]: I assigned each code to a variable at the top of the file and used those through instead of numbers

File: install_script.sh.template
Comment (randomanderson) [reply install_script.sh.template:129]: The api key is used to determine the org for the telemetry. If the api keys are different, I think in almost all cases, they'll belong to the same org.

In the extremely rare case they do belong to different orgs, I think we can live with that. The telemetry will be mostly used in aggregate across all orgs.

For upgrades, I think partial data is okay. I asked the rest of the team if we should just block telemetry on upgrades altogether instead of partial data.

File: install_script.sh.template
Comment (randomanderson) [reply install_script.sh.template:811]: The message here is quite long and spans multiple lines. I guess I could have used just the first part of the message 🤷‍♂️

Comment (randomanderson) [reply install_script.sh.template]: Actually, some of the other error messages are multiline as well. I updated to code to make it consistent

File: install_script.sh.template
Comment (randomanderson) [reply install_script.sh.template:129]: The team agrees that we shouldn't worry about the update case

**Comment by randomanderson** [comment]
@KSerrania I tested the scenarios again on production. Everything is looking good to merge once you approve.

One thing I noticed for Mac testing was that I no longer got "Unsupported Platform" because the script is invalid (mac doesn't support `declare` syntax used). To get around having blank telemetry, I grab the last couple lines of the log file if `ERROR_MESSAGE` is blank.

**Comment by olivielpeau** [comment]
@KSerrania is out this week so tagging @bkabrda for a review

**Review by bkabrda — APPROVED** [review]
LGTM now, please don't forget to remove the `DD-Telemetry-Debug-Enabled: true` header before merging.

**Pull request merged**
